poll/poll.py

```diff
@@ -300,6 +316,89 @@ def wrapper(self, request_json, suffix=''):
 
+    def _get_status(self):
+        self.check_pending_export()
+        return {
+            'export_pending': bool(self.active_export_task_id),
+            'last_export_result': self.last_export_result,
+            'download_url': self.download_url_for_last_report,
+        }
+
+    def check_pending_export(self):
+        """
+        If we're waiting for an export, see if it has finished, and if so, get the result.
+        """
+        from .tasks import export_data as export_data_task  # Import here since this is edX LMS specific
+        if self.active_export_task_id:
+            async_result = export_data_task.AsyncResult(self.active_export_task_id)
+            if async_result.ready():
+                self._store_export_result(async_result)
+
+    @property
+    def download_url_for_last_report(self):
+        """ Get the URL for the last report, if any """
+        # Unfortunately this is a bit inefficient due to the ReportStore API
+        if not self.last_export_result or self.last_export_result['error'] is not None:
+            return None
+
+        report_store = ReportStore.from_config(config_name='GRADES_DOWNLOAD')
+        course_key = getattr(self.scope_ids.usage_id, 'course_key', None)
+        return dict(report_store.links_for(course_key)).get(self.last_export_result['report_filename'])
+
+    def student_module_queryset(self):
+        return StudentModule.objects.filter(
+            course_id=self.runtime.course_id,
+            module_state_key=self.scope_ids.usage_id,
+        ).order_by('-modified')
+
+    @XBlock.json_handler
+    def csv_export(self, data, suffix=''):
+        """
+        Asynchronously export given data as a CSV file.
+        """
+        # Launch task
+        from .tasks import export_data as export_data_task  # Import here since this is edX LMS specific
+
+        # Make sure we nail down our state before sending off an asynchronous task.
+        async_result = export_data_task.delay(
+            self.prepare_data(),
+            unicode(getattr(self.runtime, 'course_id', 'course_id')),
+            self.get_filename(),
+        )
+        if not async_result.ready():
+            self.active_export_task_id = async_result.id
+        else:
+            self._store_export_result(async_result)
+
+        return self._get_status()
+
+    def _store_export_result(self, task_result):
+        """ Given an AsyncResult or EagerResult, save it. """
+        self.active_export_task_id = ''
+        if task_result.successful():
+            if isinstance(task_result.result, dict) and not task_result.result.get('error'):
+                self.last_export_result = task_result.result
+            else:
+                self.last_export_result = {'error': u'Unexpected result: {}'.format(repr(task_result.result))}
+        else:
+            self.last_export_result = {'error': unicode(task_result.result)}
+
+    @XBlock.json_handler
+    def get_status(self, data, suffix=''):
```

Comment:
Please rename this (and the related private methods and javascript code) to `get_export_status` or something less generic, and add a docstring - I would expect "get_status" to return the result of the poll or something.